lib/route/route_obj.c

```diff
@@ -145,7 +145,7 @@ static void route_dump_line(struct nl_object *a, struct nl_dump_params *p)
 		nl_dump(p, "cache ");
 
 	if (!(r->ce_mask & ROUTE_ATTR_DST) ||
-	    nl_addr_get_len(r->rt_dst) == 0)
+	    nl_addr_get_prefixlen(r->rt_dst) == 0)
```

Comment:
> Of course, a valid route can (as far as kernel is concerned) not have an non-zero host part, but still...

Actually it can't.

IPv4 will just plainly reject it if you try to add a route with a non-zero host part:

https://elixir.bootlin.com/linux/v5.18.9/source/net/ipv4/fib_trie.c#L1192

while IPv6 will just silently zero out any bits that aren't part of the prefix:

https://elixir.bootlin.com/linux/v5.18.9/source/net/ipv6/route.c#L966
https://elixir.bootlin.com/linux/v5.18.9/source/net/ipv6/route.c#L3780

So unless we have a broken kernel, all default routes will have to be all-0 in the kernel. Of course that won't stop anyone from creating "broken" `rtnl_route` objects themselves.

So if we wanted to, we could

```suggestion
	    (nl_addr_get_prefixlen(r->rt_dst) == 0  &&
	     nl_addr_get_len(r->rt_dst) > 0 && nl_addr_iszero(rt->rt_dst)))
```

That way we would print default only if the address is 0.0.0.0/0 or ::/0, and print the actual address otherwise.

The length check is necessary to avoid treating "none" as "default", as `nl_addr_iszero()` will also return true for zero length addresses (since none of the bytes are non-zero).

Reply:
> Actually it can't.

Yes, exactly my point... you cannot configure such a route in kernel. But you can create such a nl_addr instance with libnl3 (even if it's probably not gonna work).

But it would be nice to see such "broken" nl_addr in the string representation and not hide it under "default".

The suggested change lgtm!!